### PR TITLE
all: add extra information to serialized endpoints

### PR DESCRIPTION
**Pull request by josharian**
This is the big commit for which #1831 is preliminary cleanup.
We should submit it at the beginning of the 1.10 cycle.
It'll be followed shortly thereafter by another commit
removing part of our wireguard-go fork and
a few lines in magicsock{,_test}.go.

### CODE REVIEW

**Comment by josharian** [comment]
This is the commit that breaks our dependence on wireguard-go to give us the peer's public key during ParseEndpoint.

It's relatively intricate (took me the better part of a week to get right), so a skeptical review is probably warranted.

In cheerful news, with this commit, wgenginer/wuserspace.go no longer imports package net! netaddr-ification is thoroughly underway.

File: wgengine/magicsock/legacy.go [hunk 1]
```diff
@@ -34,7 +33,7 @@ var (
 	errDisabled       = errors.New("magicsock: legacy networking disabled")
 )
 
-func (c *Conn) createLegacyEndpointLocked(pk key.Public, addrs string) (conn.Endpoint, error) {
+func (c *Conn) createLegacyEndpointLocked(pk key.Public, addrs []netaddr.IPPort, rawdst string) (conn.Endpoint, error) {
```
Comment (bradfitz) [inline wgengine/magicsock/legacy.go:36]: rawDst or rawDest, but add a comment about what it is and what form it takes (comma-separated something? IPs, IP:ports?)

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -2736,42 +2736,35 @@ func packIPPort(ua netaddr.IPPort) []byte {
...
-	copy(pk[:], keyAddrs)
-	addrs := keyAddrs[len(pk):]
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	c.logf("magicsock: ParseEndpoint: key=%s: %s", pk.ShortString(), derpStr(addrs))
-
-	if !strings.HasSuffix(addrs, wgcfg.EndpointDiscoSuffix) {
-		return c.createLegacyEndpointLocked(pk, addrs)
+// endpointStr is a json-serialized wgcfg.Endpoints struct.
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2739]: not quite: it has 32 bytes at teh beginning that line 2746 strips off?

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -3115,7 +3108,7 @@ type discoEndpoint struct {
 	discoKey   tailcfg.DiscoKey // for discovery mesages
 	discoShort string           // ShortString of discoKey
 	fakeWGAddr netaddr.IPPort   // the UDP address we tell wireguard-go we're using
-	wgEndpoint string           // string from ParseEndpoint: "<hex-discovery-key>.disco.tailscale:12345"
+	wgEndpoint string           // string from ParseEndpoint, holds a serialized wgcfg.Endpoints
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:3111]: JSON-serialized

File: wgengine/wgcfg/config.go [hunk 4]
```diff
@@ -31,10 +29,59 @@ type Config struct {
...
 }
 
+// Endpoints represents the routes to reach a remote node.
+// It is serialized and provided to wireguard-go as a conn.Endpoint.
+type Endpoints struct {
+	// PublicKey is the public key for the remote node.
+	PublicKey wgkey.Key `json:"pk"`
+	// DiscoKey is the disco key associated with the remote node.
+	DiscoKey tailcfg.DiscoKey `json:"dk,omitempty"`
+	// IPPorts is a set of possible ip+ports the remote node can be reached at.
```
Comment (bradfitz) [inline wgengine/wgcfg/config.go:43]: this is only used for legacy endpoints, right? add docs?

File: wgengine/wgcfg/config.go [hunk 5]
```diff
@@ -31,10 +29,59 @@ type Config struct {
...
+	IPPorts []netaddr.IPPort `json:"ipp,omitempty"`
```
Comment (bradfitz) [inline wgengine/wgcfg/config.go:44]: slice in a value type is a little scary. normally the rule is no value types with mutable pointers.

can we use some immutable IPPortSet type instead? Or flatten it to a comma-separated strings with a method to get them back out parsed?

File: wgengine/magicsock/legacy.go (hunk 1 above)
Comment (josharian) [reply wgengine/magicsock/legacy.go:36]: Done. (The answer is: don't look inside it. Just return it from addrSet.DstToString.)

File: wgengine/magicsock/magicsock.go (hunk 2 above)
Comment (josharian) [reply wgengine/magicsock/magicsock.go:2739]: That's temporary. Added a parenthetical comment, which I'll delete shortly.

File: wgengine/wgcfg/config.go [hunk 6]
```diff
@@ -31,10 +30,92 @@ type Config struct {
...
 
+// Endpoints represents the routes to reach a remote node.
+// It is serialized and provided to wireguard-go as a conn.Endpoint.
+type Endpoints struct {
+	// PublicKey is the public key for the remote node.
+	PublicKey wgkey.Key `json:"pk"`
+	// DiscoKey is the disco key associated with the remote node.
+	DiscoKey tailcfg.DiscoKey `json:"dk,omitempty"`
+	// IPPorts is a set of possible ip+ports the remote node can be reached at.
+	// This is used only for legacy connections.
```
Comment (bradfitz) [inline wgengine/wgcfg/config.go:45]: legacy connections to pre-disco (pre-0.100) peers.

**Review by bradfitz — APPROVED** [review]
Maybe file tracking bug for FromUAPI optimizations in general? Lot of stuff here.

**Comment by josharian** [comment]
Filed https://github.com/tailscale/tailscale/issues/1912 for optimizing FromUAPI.

**Pull request merged**